File: src/sungrowlib/transports/HttpTransport.py
"""
WiNet HTTP connection for Sungrow inverters.

In detail this will query the inverter via websocket for the token and then
query the inverter via http for the actual data.

Data-queries via websocket are not yet implemented.
"""

import asyncio
import logging
import os
import time
from enum import Enum
from typing import TypeVar, cast

import aiohttp

from sungrowlib.types import (
    ConnectionError,
    GenericError,
    InvalidResponseError,
    RegisterRange,
    RegisterType,
    SungrowlibRuntimeError,
)
from sungrowlib.util import typesafe_get

logger = logging.getLogger(__name__)


T = TypeVar("T")


class HttpTransport:
    def __init__(self, host: str, port: int | None = None):
        if not port:
            port = self.default_port()
        self._host = host
        self._port = port

        # TODO: document why we keep an aiohttp.ClientSession around
        timeout = aiohttp.ClientTimeout(total=10)
        self._aio_client = aiohttp.ClientSession(timeout=timeout)
        self._ws: aiohttp.client.ClientWebSocketResponse | None = None

        self._token: str | None = None
        self._inverter: dict[str, str] | None = None

    @staticmethod
    def default_port() -> int:
        return 8082

    async def _ws_query(
        self, query: dict[str, str | int], return_key: str, return_type: type[T]
    ) -> T:
        # Potential services: connect, devicelist, state, statistics, runtime, real
        if self._ws is None:
            raise SungrowlibRuntimeError("Websocket not connected")

        try:
            await self._ws.send_json(query)
            response: dict[str, object] = await self._ws.receive_json()
        except Exception as e:
            logger.error(f"Cannot send message to inverter: {e}")
            await self.disconnect()
            raise ConnectionError(e)

        if response.get("result_code") != 1 or response.get("result_msg") != "success":
            raise InvalidResponseError(f"Invalid websocket response: {response}")

        result_data = typesafe_get(response, "result_data", dict[str, object])
        return typesafe_get(result_data, return_key, return_type)

    async def _get_new_token(self) -> str:
        assert self._token is None

        return await self._ws_query(
            {"lang": "en_us", "token": "", "service": "connect"}, "token", str
        )

    def _debug(self, msg: str):
        logger.debug(f"{self._host}:{self._port} {msg}")

    async def _get_connected_devices(
        self,
    ):
        assert self._token is not None

        return await self._ws_query(
            {
                "lang": "en_us",
                "token": self._token,
                "service": "devicelist",
                "type": "0",
                "is_check_token": "0",
            },
            "list",
            list[dict[str, str]],
        )

    async def connect(self):
        """
        Establish permanent websocket connection to the WiNet dongle
        and retrieve a token.

        Raises ConnectionError if the connection fails.
        """

        if self.connected:
            return

        self._debug("Connecting...")

        ws_endpoint = f"ws://{self._host}:{self._port}/ws/home/overview"
        try:
            self._ws = await self._aio_client.ws_connect(ws_endpoint)

            self._token = await self._get_new_token()

            # The first device is always the inverter.
            # TODO: can we do anything with the others?
            devices = await self._get_connected_devices()
            self._debug(f"Connected devices: {devices}")
            if not devices:
                raise InvalidResponseError(
                    "No devices connected to inverter. This is unexpected."
                )

            self._inverter = devices[0]
            self._debug("Connection via websocket established")
        except ConnectionError:
            await self.disconnect()
            raise
        except TimeoutError as e:
            await self.disconnect()
            raise ConnectionError("Timeout while connecting to inverter") from e
        except Exception as e:
            await self.disconnect()
            raise ConnectionError() from e

    async def disconnect(self):
        self._debug("Disconnecting...")

        self._token = None
        self._inverter = None

        if self._ws:
            await self._ws.close()
            self._ws = None

        # Force completely new connections on next connect()
        await self._aio_client.close()

    @property
    def connected(self) -> bool:
        return self._token is not None

    async def _query_GET(self, url: str, params: dict[str, str | int]):
        """Query the inverter via http GET."""

        try:
            async with await self._aio_client.get(url, params=params) as r:
                if r.status == 200:
                    return cast(dict[str, object], await r.json())
                else:
                    logger.error(
                        "Invalid response from inverter: %s %s", r.status, r.text
                    )
                    raise InvalidResponseError()
        except Exception as e:
            # e.g. response is not valid json
            self._debug(f"Cannot query inverter: {e}")
            raise

    def _build_http_request_for_register_query(
        self, rr: RegisterRange
    ) -> tuple[str, dict[str, str | int]]:
        assert self._inverter
        assert self._token

        param_type = {
            RegisterType.READ: 0,
            RegisterType.HOLD: 1,
        }[rr.register_type]

        # Usually port 80, but we cannot use a hardcoded port in tests.
        # In tests we'll simply reuse the port of the websocket server.
        if "PYTEST_CURRENT_TEST" in os.environ:
            logger.warning("Running in test mode, using websocket port for http")
            port = self._port
        else:
            port = 80

        url = f"http://{self._host}:{port}/device/getParam"
        params: dict[str, str | int] = {
            "dev_id": self._inverter["dev_id"],
            "dev_type": self._inverter["dev_type"],
            "dev_code": self._inverter["dev_code"],
            "type": "3",  # TODO: Why 3?
            "param_addr": rr.start,
            "param_num": rr.length,
            "param_type": param_type,
            "token": self._token,
            "lang": "en_us",
            "time123456": int(time.time()),
        }
        return url, params

    async def _query_http_json(
        self, rr: RegisterRange, retries: int = 3
    ) -> dict[object, object]:
        for attempt in range(retries):
            if attempt > 0:
                # Add 1 second delay before next attempt
                await asyncio.sleep(1)

            try:
                await self.connect()

                # (Re-)build query with current token
                url, params = self._build_http_request_for_register_query(rr)

                response = await self._query_GET(url, params)

                code = _parse_inverter_result_code(response)
                if code == InverterResultCode.SUCCESS:
                    data = response["result_data"]
                    if not isinstance(data, dict):
                        logger.error(
                            f"Invalid response from inverter: {response}, "
                            f"result_data is not a dictionary"
                        )
                        raise InvalidResponseError("Invalid response from inverter")
                    return cast(dict[object, object], data)
                elif code == InverterResultCode.TOKEN_EXPIRED:
                    logger.debug("Token expired, reconnecting")
                    await self.disconnect()
                elif code == InverterResultCode.BUSY:
                    logger.debug("Inverter busy, will retry after some delay")
                    await asyncio.sleep(5)
                else:
                    logger.error(
                        f"Invalid response from inverter: {response}, "
                        f"result_code: {code}"
                    )
                    raise InvalidResponseError("Invalid response from inverter")
            except GenericError as e:
                logger.debug(f"Cannot connect to inverter: {e}")

        # Append last_error / all errors to exception?!
        raise ConnectionError("Repeatedly failed to query inverter")

    async def read_range(self, register_range: RegisterRange) -> list[int]:
        # Note: websocket does not allow access to all possible registers.
        # Not quite clear whether it's worth the effort to query some via websocket and
        # only the rest via http. Potentially better error messages??

        json = await self._query_http_json(register_range)
        json = cast(dict[str, str], json)
        return _parse_modbus_data(json, register_range.length)

    def __str__(self):
        return f"http({self._host}:{self._port})"


class InverterResultCode(Enum):
    SUCCESS = 1
    TOKEN_EXPIRED = 106
    BUSY = 301


def _parse_inverter_result_code(response: dict[str, object]):
    logger.debug(f"Response: {response}")

    try:
        return InverterResultCode(response["result_code"])
    except KeyError:
        raise InvalidResponseError(
            f"Invalid response from inverter (no result_code: {response})"
        )
    except ValueError:
        raise InvalidResponseError(
            f"Invalid response from inverter (invalid result_code: {response})"
        )


def _parse_modbus_data(response_json: dict[str, str], expected_length: int):
    modbus_data = response_json["param_value"].split(" ")
    logger.debug(f"Got modbus data: {modbus_data}")

    # There is always an extra null at the end, remove it.
    modbus_data.pop()

    if len(modbus_data) != expected_length * 2:
        logger.error(
            "Invalid response from inverter: "
            f"{response_json} => {modbus_data}, "
            f"expected length {expected_length}"
        )
        raise InvalidResponseError("Incorrect length of modbus data")

    # Merge two consecutive bytes into 16 bit integers, same as pymodbus.
    # Maybe it would be better to use bytes everywhere...
    # but pymodbus was implemented first.
    data: list[int] = []
    for i in range(0, len(modbus_data), 2):
        data.append(int(modbus_data[i], 16) * 256 + int(modbus_data[i + 1], 16))  # noqa: PERF401
    return data

File: ts/src/transport/http-transport.ts
import type { Transport } from './transport.js';
import {
  ConnectionError, InvalidResponseError,
  BusyError, TokenExpiredError, TooManyRetriesError,
} from '../core/errors.js';

export interface WiNetDevice {
  dev_id: number;
  dev_type: number;
  dev_code: number;
}

export interface WsMessage {
  result_code: number;
  result_msg: string;
  result_data: unknown;
}

export interface HttpResponse {
  result_code: number;
  result_msg: string;
  result_data: { param_value?: string } | null;
}

export interface HttpTransportDeps {
  openWebSocket(url: string): Promise<WebSocketLike>;
  httpGet(url: string): Promise<HttpResponse>;
}

export interface WebSocketLike {
  send(data: string): void;
  close(): void;
  onMessage(handler: (data: string) => void): void;
}

export interface HttpTransportOptions {
  host: string;
  httpPort?: number;
  wsPort?: number;
  maxRetries?: number;
  retryDelayMs?: number;
  busyDelayMs?: number;
  deps: HttpTransportDeps;
}

const DEFAULT_HTTP_PORT = 80;
const DEFAULT_WS_PORT = 8082;
const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_RETRY_DELAY_MS = 1000;
const DEFAULT_BUSY_DELAY_MS = 5000;

export function parseHexRegisters(hexString: string): number[] {
  const bytes = hexString.trim().split(/\s+/).map((h) => parseInt(h, 16));
  if (bytes.length > 0 && bytes[bytes.length - 1] === 0) {
    bytes.pop();
  }
  const registers: number[] = [];
  for (let i = 0; i + 1 < bytes.length; i += 2) {
    registers.push((bytes[i] << 8) | bytes[i + 1]);
  }
  return registers;
}

export function buildRegisterMap(start: number, registers: number[]): Map<number, number> {
  const map = new Map<number, number>();
  for (let i = 0; i < registers.length; i++) {
    map.set(start + i, registers[i]);
  }
  return map;
}

export class HttpTransport implements Transport {
  private _host: string;
  private _httpPort: number;
  private _wsPort: number;
  private _maxRetries: number;
  private _retryDelayMs: number;
  private _busyDelayMs: number;
  private _deps: HttpTransportDeps;

  private _token: string | null = null;
  private _device: WiNetDevice | null = null;
  private _ws: WebSocketLike | null = null;
  private _slaveId = 1;

  constructor(options: HttpTransportOptions) {
    this._host = options.host;
    this._httpPort = options.httpPort ?? DEFAULT_HTTP_PORT;
    this._wsPort = options.wsPort ?? DEFAULT_WS_PORT;
    this._maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
    this._retryDelayMs = options.retryDelayMs ?? DEFAULT_RETRY_DELAY_MS;
    this._busyDelayMs = options.busyDelayMs ?? DEFAULT_BUSY_DELAY_MS;
    this._deps = options.deps;
  }

  get connected(): boolean { return this._token !== null; }
  get token(): string | null { return this._token; }
  get device(): WiNetDevice | null { return this._device; }

  async connect(): Promise<void> {
    const wsUrl = `ws://${this._host}:${this._wsPort}/ws/home/overview`;
    try {
      this._ws = await this._deps.openWebSocket(wsUrl);
    } catch (err) {
      throw new ConnectionError(`WebSocket connect to ${wsUrl} failed`, {
        cause: err instanceof Error ? err : undefined,
      });
    }

    this._token = await this.authenticate();
    this._device = await this.discoverDevice();
  }

  async disconnect(): Promise<void> {
    if (this._ws) {
      this._ws.close();
      this._ws = null;
    }
    this._token = null;
    this._device = null;
  }

  setSlaveId(id: number): void {
    this._slaveId = id;
  }

  async readInputRegisters(start: number, count: number): Promise<Map<number, number>> {
    return this.readRegisters('input', start, count);
  }

  async readHoldingRegisters(start: number, count: number): Promise<Map<number, number>> {
    return this.readRegisters('hold', start, count);
  }

  private async readRegisters(type: 'input' | 'hold', start: number, count: number): Promise<Map<number, number>> {
    if (!this._token || !this._device) {
      throw new ConnectionError('Not connected — call connect() first');
    }

    for (let attempt = 0; attempt < this._maxRetries; attempt++) {
      const params = new URLSearchParams({
        token: this._token,
        dev_id: String(this._device.dev_id),
        dev_type: String(this._device.dev_type),
        dev_code: String(this._device.dev_code),
        type: type === 'hold' ? '1' : '0',
        param_addr: String(start),
        param_num: String(count),
      });

      const url = `http://${this._host}:${this._httpPort}/device/getParam?${params}`;
      let resp: HttpResponse;

      try {
        resp = await this._deps.httpGet(url);
      } catch (err) {
        throw new ConnectionError(`HTTP request failed: ${url}`, {
          cause: err instanceof Error ? err : undefined,
        });
      }

      if (resp.result_code === 1) {
        const paramValue = resp.result_data?.param_value;
        if (paramValue === undefined || paramValue === null) {
          throw new InvalidResponseError('Missing param_value in response');
        }
        const registers = parseHexRegisters(paramValue);
        return buildRegisterMap(start, registers);
      }

      if (resp.result_code === 106) {
        this._token = null;
        throw new TokenExpiredError();
      }

      if (resp.result_code === 301) {
        if (attempt < this._maxRetries - 1) {
          await delay(this._busyDelayMs);
          continue;
        }
        throw new BusyError();
      }

      throw new InvalidResponseError(`Unexpected result_code ${resp.result_code}: ${resp.result_msg}`);
    }

    throw new TooManyRetriesError(this._maxRetries);
  }

  private authenticate(): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      if (!this._ws) { reject(new ConnectionError('WebSocket not open')); return; }

      const timeout = setTimeout(() => reject(new ConnectionError('WebSocket auth timed out')), 10000);

      this._ws.onMessage((raw) => {
        clearTimeout(timeout);
        try {
          const msg: WsMessage = JSON.parse(raw);
          if (msg.result_code !== 0) {
            reject(new ConnectionError(`Auth failed: ${msg.result_msg}`));
            return;
          }
          const data = msg.result_data as { token?: string };
          if (!data?.token) {
            reject(new InvalidResponseError('No token in auth response'));
            return;
          }
          resolve(data.token);
        } catch (err) {
          reject(new InvalidResponseError(`Invalid auth response: ${raw}`));
        }
      });

      this._ws.send(JSON.stringify({ service: 'connect', token: '' }));
    });
  }

  private discoverDevice(): Promise<WiNetDevice> {
    return new Promise<WiNetDevice>((resolve, reject) => {
      if (!this._ws) { reject(new ConnectionError('WebSocket not open')); return; }

      const timeout = setTimeout(() => reject(new ConnectionError('Device discovery timed out')), 10000);

      this._ws.onMessage((raw) => {
        clearTimeout(timeout);
        try {
          const msg: WsMessage = JSON.parse(raw);
          if (msg.result_code !== 0) {
            reject(new ConnectionError(`Device discovery failed: ${msg.result_msg}`));
            return;
          }
          const devices = msg.result_data as WiNetDevice[];
          if (!Array.isArray(devices) || devices.length === 0) {
            reject(new InvalidResponseError('No devices found'));
            return;
          }
          resolve(devices[0]);
        } catch {
          reject(new InvalidResponseError(`Invalid device list response: ${raw}`));
        }
      });

      this._ws.send(JSON.stringify({ service: 'devicelist' }));
    });
  }
}

function delay(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
